### PR TITLE
Add period beginning feature

### PR DESCRIPTION
**Pull request by rajeee**
## Pull Request Description

Currently, OS-HPXML outputs timestamps in period ending format. This PR enables converting them to period beginning format. We require period beginning for LA100ES project. 

## Checklist

Not all may apply

- [x] Code changes (must work)
- [ ] Tests exercising your feature/bug fix (check coverage report on Checks -> BuildStockBatch Tests -> Artifacts)
- [x] Coverage has increased or at least not decreased. Update `minimum_coverage` in `.github/workflows/ci.yml` as necessary.
- [x] All other unit tests passing
- [x] Update validation for project config yaml file changes
- [x] Update existing documentation
- [x] Run a small batch run to make sure it all works (local is fine, unless an Eagle specific feature)
- [x] Add to the changelog_dev.rst file and propose migration text in the pull request

### CODE REVIEW

**Comment by joseph-robertson** [comment]
I wonder if this would be more efficiently handled (optionally) on the OS-HPXML side...

**Comment by shorowit** [comment]
To be clear, OS-HPXML just outputs whatever EnergyPlus outputs, which is indeed period ending timestamps. I assume outputs from ComStock are the same. It seems like it could cause confusion to stray from the EnergyPlus convention...?

File: buildstockbatch/base.py [hunk 1]
```diff
@@ -164,7 +165,23 @@ def make_sim_dir(building_id, upgrade_idx, base_dir, overwrite_existing=False):
...
 
     @staticmethod
-    def cleanup_sim_dir(sim_dir, dest_fs, simout_ts_dir, upgrade_id, building_id):
+    def make_period_beginning(df):
+        step = df['time'].iloc[1] - df['time'].iloc[0]
+        if 'timedst' in df.columns:
+            diff = df['timedst'] - df['time']
+            offset_diff = np.array([0] + list(diff.values)[:-1])
+            df['time'] = df['time'] - step
+            df['timedst'] = df['time'] + np.array(offset_diff)
```
Comment (afontani) [inline buildstockbatch/base.py]: L174 is a good solution to the daylight savings shifting in timestamps.

File: buildstockbatch/schemas/v0.3.yaml [hunk 2]
```diff
@@ -109,6 +109,7 @@ postprocessing-spec:
   partition_columns: list(str(), required=False)
   aws: include('aws-postprocessing-spec', required=False)
   keep_individual_timeseries: bool(required=False)
+  timestamps_uses_period_beginning: bool(required=False)
```
Comment (afontani) [inline buildstockbatch/schemas/v0.3.yaml]: Just to be clear, is this the proper input in the YAML?
```
postprocessing:
  timestamps_uses_period_beginning: true
  aws:
    ...
```

**Review by afontani — COMMENTED** [review]
@rajeee The implementation of the feature makes sense to me. I will leave it up to @nmerket if this warrants a merge with `develop`. If we want to merge this into `develop`, we should write some tests making sure that the shifting is done correctly especially around daylight savings times and the end of year timestamp.

**Comment by afontani** [comment]
> To be clear, OS-HPXML just outputs whatever EnergyPlus outputs, which is indeed period ending timesteps. I assume outputs from ComStock are the same. It seems like it could cause confusion to stray from the EnergyPlus convention...?

@shorowit, it could definitely get confusing. This capability has been wanted when we connect with other tools and teams. Examples are the LA100, LA100 ES, and dsgrid projects. In these cases, we as a larger group of modeling teams have to come to a consensus about either period ending or period beginning timestamps. In these cases, we have been forced to post-process our data to conform to the needs of other teams (shifting our energy timeseries to period beginning timestamps).

File: buildstockbatch/schemas/v0.3.yaml (hunk 2 above)
Comment (rajeee) [reply buildstockbatch/schemas/v0.3.yaml]: Yes.

**Comment by afontani** [comment]
I did a comparison between the timestamps before and after with `buildstock_docker`. I used 2012 as a comparison point. The daylight savings dates for 2012 are March 11 - November 4.
timestep-comparison
.

**Comment by rajeee** [comment]
@nmerket @shorowit 
What do you think of merging in this feature to develop? If you think it is worthwhile to merge this to develop, I will spend some time to write tests for this. If you think this feature is too project specific, I will not bother.

**Comment by nmerket** [comment]
I think period beginning timestamps are better for a variety of reasons. However, I do have concerns about unanticipated downstream implications of this change. I like having it available to anyone as long as its optional and turned off by default. But yes, I do think some tests are in order as there are a lot of weird edge cases with timestamps.

File: buildstockbatch/base.py [hunk 3]
```diff
@@ -227,6 +244,8 @@ def get_clean_column_name(x):
                 return x.lower()
 
             tsdf.rename(columns=get_clean_column_name, inplace=True)
+            if cfg.get("postprocessing", {}).get("timestamps_use_period_beginning", False):
+                BuildStockBatchBase.make_period_beginning(tsdf)
```
Comment (nmerket) [inline buildstockbatch/base.py]: Might be worth turning this into a [classmethod](https://docs.python.org/3/library/functions.html#classmethod) then you can implicitly reference the class.

**Review by nmerket — COMMENTED** [review]
Looks pretty good. One minor comment below. 

One more thought: It would be nice to have a test exercising the `make_period_beginning` method.

**Comment by shorowit** [comment]
My opinion: I would still rather see this as an OS-HPXML feature than a BSB feature. Interest in this capability is not specific to ResStock, so it would be nice to make it available to a wider set of users. And it would address [this issue](https://github.com/NREL/OpenStudio-HPXML/issues/721).

**Comment by shorowit** [comment]
Though I just realized that implementing it in OS-HPXML might only benefit ResStock (and other residential tools), but not ComStock.

**Comment by nmerket** [comment]
All good points, @shorowit. I say we just go with this since it's already done and meets the immediate need. You can add it to OpenStudio-HPXML later to support other residential tools. We just need to make sure to not turn it on twice.

**Review by nmerket — CHANGES_REQUESTED** [review]
This is probably ready to merge except that it needs some testing coverage of the new feature.

**Comment by afontani** [comment]
I might suggest closing this and using #311

**Comment by joseph-robertson** [comment]
> I might suggest closing this and using #311

If period begin timestamps are needed for ComStock, we'd probably want/need to use this PR for ComStock and #311 for ResStock.

**Comment by afontani** [comment]
@joseph-robertson: Good point. 

@rHorsey , or @asparke2 : Do you want the ability to automatically convert ComStock timeseries from period ending to period beginning? There is a path for OpenStudio-HPXML to automatically do this, but @rajeee did this separately through post-processing the data. Any preferences?

**Comment by asparke2** [comment]
I would strongly prefer that buildstockbatch always reports out a consistent (either beginning or ending) format, and that any users perform postprocessing after the fact to their preferred format. As soon as we put this switch in, we are going to have to remember to track which run used which setting in buildstockbatch, which will be impossible after more than a few months.  If you remember back to LA100, we had tons of problems related to how various runs were dealing with leap years and daylight savings time, which I'd really prefer not to relive.  Therefore, I would strongly prefer NOT to merge this feature.

**Comment by afontani** [comment]
@asparke2 : Thanks for chiming in. For ResStock this feature is in response to OS-HPXML changing its default to period beginning timestamps (https://github.com/NREL/OpenStudio-HPXML/pull/1198), so ResStock needs some feature to keep the simulations consistent with period ending timestamps.

I have also suggested we save the YAML alongside the results in this issue to capture the exact simulation of the project (#254) which might help the remembering part of your comment.

I guess I still suggest closing this PR as ComStock will likely not use this feature.

**Comment by asparke2** [comment]
It seems like instead of in buildstockbatch, the appropriate place for this code is then inside the Measure that ResStock uses to pull the timeseries results out of the E+ input file. I do like the idea of uploading the YML (and also the builstock.csv) to S3 for easier debugging for posterity though.

**Comment by afontani** [comment]
@rajeee : This PR was just merged, https://github.com/NREL/buildstockbatch/pull/311. 

Are you okay with closing this PR? We might want to keep this branch for now until LA100 ES wraps up.
